sympy/functions/elementary/trigonometric.py

```diff
@@ -107,20 +107,22 @@ def _peeloff_pi(arg):
     (x + pi*y + pi/6, pi/2)
 
     """
+    pi_coeff = S.Zero
     for a in Add.make_args(arg):
         if a is S.Pi:
-            K = S.One
+            pi_coeff = S.One
             break
         elif a.is_Mul:
-            K, p = a.as_two_terms()
-            if p is S.Pi and K.is_Rational:
-                break
+            K = a.coeff(S.Pi)
+            if K and K.is_rational:
+                pi_coeff += K
     else:
-        return arg, S.Zero
+        if pi_coeff is S.Zero:
+            return arg, S.Zero
 
-    m1 = (K % S.Half)*S.Pi
-    m2 = K*S.Pi - m1
-    return arg - m2, m2
+    m1 = (pi_coeff % S.Half)*S.Pi
+    m2 = pi_coeff*S.Pi - m1
+    return (arg - m2).cancel(), m2
```

Comment:
Using `cancel` is potentially very expensive.  We should not use this as part of normal evaluation.

Reply:
See some simplification is required to `cancel` out the terms. `expand()` might work, but I think `cancel()` is more effective.

Reply:
Both `expand` and `cancel` can be very slow for larger expressions. Unless there is some way to know that this is always a small expression then we should not use either function. The function is written the way it is now is to try and keep it efficient. If we can extend its capability without risking a slowdown then that's fine.

If instead this was handled as a simplification function then it could be okay to try slower operations but this routine is called every time we have `sin(expr)` etc.

Reply:
These are the type of expressions:

```
In [5]: y
Out[5]: pi*m + 2*pi*n - pi*(m + 2*n)

In [6]: y.cancel()
Out[6]: 0

In [7]: y.expand()
Out[7]: 0

In [11]: y
Out[11]:
       pi*(m - 1)      /m       1\
pi*n + ---------- - pi*|- + n - -| + 1
           2           \2       2/

In [12]: y.expand()
Out[12]: 1

In [13]: y.cancel()
Out[13]: 1
```

Though `expand` is working, but I am not sure if it will work if the expressions get too complicated. That's the reason, I chose `cancel`.

Reply:
> If instead this was handled as a simplification function then it could be okay to try slower operations but this routine is called every time we have sin(expr) etc.

I totally agree with you, do we have faster simplification functions for these type of expressions ?